File: 2.add-two-numbers/solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        val sentinel = ListNode(0)
        var current = sentinel
        var p1 = l1
        var p2 = l2
        var carry = 0

        while (p1 != null || p2 != null || carry > 0) {
            var sum = carry
            if (p1 != null) {
                sum += p1.`val`
                p1 = p1.next
            }
            if (p2 != null) {
                sum += p2.`val`
                p2 = p2.next
            }
            current.next = ListNode(sum % 10)
            current = current.next
            carry = sum / 10
        }

        return sentinel.next
    }
}

File: 2.add-two-numbers/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    sentinel := &ListNode{}
    current := sentinel
    carry := 0

    for l1 != nil || l2 != nil || carry > 0 {
        v1 := 0
        if l1 != nil {
            v1 = l1.Val
            l1 = l1.Next
        }

        v2 := 0
        if l2 != nil {
            v2 = l2.Val
            l2 = l2.Next
        }
        
        sum := v1 + v2 + carry
        carry = sum / 10
        current.Next = &ListNode{Val: sum % 10}
        current = current.Next
    }

    return sentinel.Next
}